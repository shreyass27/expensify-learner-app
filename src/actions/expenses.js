import uuid from 'uuid';
import dataBase from '../firebase/firebase';

const handleError = (error) => console.log('Error Occured', error)

const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt  = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        return dataBase.ref('expenses').push(expense)
            .then(
                (snapShot) => {
                    console.log(snapShot);
                    dispatch(addExpense({
                        id: snapShot.key,
                        ...expense
                    }));
                }
            )
            .catch(handleError);

    }

}


export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});



export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return dataBase.ref(`expenses/${id}`).update(updates)
            .then(
                (snapShot) => {
                    console.log(snapShot);
                    dispatch(editExpense(id, updates));
                }
            )
            .catch(handleError);

    }

}

const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id }) => {
    return ( dispatch ) => {
        return dataBase.ref(`expenses/${id}`).remove()
            .then(
                () => dispatch(removeExpense({id}))
            )
            .catch(handleError)
    }
}

const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return dataBase.ref('expenses').once('value')
            .then( (snapShot) => {
                const expenses = [];
                snapShot.forEach((expense) => {
                    expenses.push({
                        id: expense.key,
                        ...expense.val()
                    })
                });
                dispatch(setExpenses(expenses));
            })
            .catch(handleError);
    }
}
