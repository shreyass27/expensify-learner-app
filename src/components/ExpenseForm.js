import React from 'react';
import momemt from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount).toString() : '',
            createdAt:  props.expense ? momemt(props.expense.createdAt) : momemt(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(
            () => ({ description })
        )
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(
            () => ({ note })
        )
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(
                () => ({ amount })
            )
        }
    }

    onDateChange = (createdAt) => {
        if ( createdAt ) {
            this.setState({ createdAt });
        }
    }
    onFocusChange = ({ focused }) => this.setState({ calendarFocused: focused });

    onSubmit = (e) => {
        e.preventDefault();
        if ( !this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please add a description and amount'}));
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: +(this.state.amount),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder="Description" autoFocus 
                        value={this.state.description} onChange={this.onDescriptionChange}/>
                    <input type="text " placeholder="Amount"
                    value={this.state.amount} onChange={this.onAmountChange} />
                    <br />
                    <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={this.onDateChange} // PropTypes.func.isRequired
                    focused={this.state.calendarFocused} // PropTypes.bool
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                    id="createdAtCalendar" // PropTypes.string.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                    />
                    <textarea
                        placeholder="Please add a note if required"
                        value={this.state.note} onChange={this.onNoteChange}
                    >

                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}
