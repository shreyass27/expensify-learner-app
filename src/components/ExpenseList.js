import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses'
import ExpenseItemList from './ExpenseListItem';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p> No expense added yet.</p>
            ) : 
            (props.expenses.map(
                expense => <ExpenseItemList key={expense.id} {...expense} />
            ))
            
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
