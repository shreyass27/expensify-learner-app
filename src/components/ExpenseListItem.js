import React from 'react';
import { Link } from  'react-router-dom';
import moment from 'moment';

const ExpenseItemList = ({description, amount, createdAt, id, dispatch}) => (
    <div>
        <Link to={`/edit/${id}`}><h3>{description} </h3></Link>
        <p>{amount} - {moment(createdAt).format('Do MMM, YYYY')}</p>
    </div>
);

export default ExpenseItemList;