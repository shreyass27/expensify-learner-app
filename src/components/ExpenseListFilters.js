import React from 'react';
import { connect } from 'react-redux';
import { setFilterText, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = focusedInput => this.setState({ calendarFocused: focusedInput });


    onTextChange = (e)=> {
        this.props.setFilterText(e.target.value);
    }

    onSortChange = (e)=> {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }
    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                    showClearDates={true}
                />
            </div>
        );
    }
}

const mapStatetoProps = (state) => {
    return {
        filters: state.filters
    };
};
const mapDispatchToProps = (dispatch) => ({
    setFilterText: (text) => dispatch(setFilterText(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})
export default connect(mapStatetoProps, mapDispatchToProps)(ExpenseListFilters);