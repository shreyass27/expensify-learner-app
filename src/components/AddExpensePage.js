import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import  { startAddExpense } from '../actions/expenses';

const AddExpensePage = (props) => (
    <div>
        <h2>Add Expense</h2>
        <ExpenseForm 
            onSubmit={
                (expense) => {
                    props.dispatch(startAddExpense(expense))
                    props.history.push('/');
                    console.log('AddExpensePage onSubmit', expense);
                }
            }
        />
    </div>
);

export default connect()(AddExpensePage);