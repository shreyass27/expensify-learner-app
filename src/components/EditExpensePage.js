import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from  './ExpenseForm';
import { editExpense } from '../actions/expenses';
import { removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
    return (
        <div>
           <h4>Edit Expense: {props.expense.description}</h4> 
        <ExpenseForm 
            expense={props.expense}
            onSubmit={
                (expense) => {
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/');
                    console.log('EditExpensePage onSubmit', expense);
                }
            }
        />
        <button onClick={() => {
            console.log('EditExpensePage removeExpense', props.expense);
            props.dispatch(removeExpense({id: props.expense.id}));
            props.history.push('/');
        }}>Remove</button>
        </div>
    );

}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(expense => props.match.params.id === expense.id)
    }
};

export default connect(mapStateToProps)(EditExpensePage);