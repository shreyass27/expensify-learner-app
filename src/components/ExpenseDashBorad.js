import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters'

const ExpenseDashBorad = (props) => (
    <div>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashBorad;