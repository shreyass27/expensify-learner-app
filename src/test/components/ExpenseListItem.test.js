import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseItemList } from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';

test('shoould render expense item with given expense data', () => {
    const expense = expenses[2];
    const wrapper = shallow(<ExpenseItemList {...expense} />);

    expect(wrapper).toMatchSnapshot();
})