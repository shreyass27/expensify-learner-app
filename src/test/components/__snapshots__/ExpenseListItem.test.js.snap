// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shoould render expense item with given expense data 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/3"
  >
    <h3>
      Credit Card
       
    </h3>
  </Link>
  <p>
    $45.00
    -
    5th Jan, 1970
  </p>
</div>
`;
